### PR TITLE
Ported code generator to the new mini-table build API

### PR DESCRIPTION
**Pull request by haberman**
This change swaps out the implementation of mini-table building from the old code in `upbc/message_layout.{h,cc}` to the new mini-table building API in `upb/mini_table.h`.

The other code-path that builds mini-tables for types loaded at runtime (located in `upb/def.c`) is not affected. It will be ported next, so that we have only a single code path for constructing layouts and mini-tables.

The compiler should probably be refactored so that `protoc-gen-upb.cc` is not so large, but this can come in a follow-up PR. We will soon want to offer richer APIs for other code generators to use; for example, an API to return an encoded mini-descriptor, so that another code generator can embed it into another language's generated code (eg. Dart).

### CODE REVIEW

File: upb/mini_table.c [hunk 1]
```diff
@@ -418,22 +470,26 @@ static void upb_MiniTable_SetField(upb_MtDecoder* d, uint8_t ch,
...
-  field->descriptortype = kUpb_EncodedToType[type];
-  if (upb_MiniTable_HasSub(type, msg_modifiers)) {
-    field->submsg_index = sub_count ? (*sub_count)++ : 0;
-  }
+  upb_MiniTable_SetTypeAndSub(field, kUpb_EncodedToType[type], sub_count,
+                              msg_modifiers);
 }
 
-static void upb_MtDecoder_ModifyField(upb_MtDecoder* d, uint32_t mod,
+#include <stdio.h>
```
Comment (fowles) [inline upb/mini_table.c:477]: debuging aid?

File: upb/mini_table.c [hunk 2]
```diff
@@ -479,6 +537,42 @@ static void upb_MtDecoder_PushOneof(upb_MtDecoder* d, upb_LayoutItem item) {
...
+}
+
+size_t upb_MtDecoder_AlignOfRep(upb_FieldRep rep,
+                                upb_MiniTablePlatform platform) {
+  static const uint8_t kRepToAlign32[] = {
+      [kUpb_FieldRep_1Byte] = 1,   [kUpb_FieldRep_4Byte] = 4,
+      [kUpb_FieldRep_Pointer] = 4, [kUpb_FieldRep_StringView] = 4,
+      [kUpb_FieldRep_8Byte] = 8,
+  };
+  static const uint8_t kRepToAlign64[] = {
```
Comment (fowles) [inline upb/mini_table.c:565]: here and the above function, I would rather have done this with a single table that uses `UPB_SIZE` to assign values.  Is that not possible for some reason?

File: upb/mini_table.c [hunk 3]
```diff
@@ -884,34 +962,71 @@ upb_MiniTable* upb_MiniTable_BuildMapEntry(upb_FieldType key_type,
...
+    uint32_t val;
+    memcpy(&val, ptr, 4);
+    if (val >= 64) break;
+    table->mask |= 1ULL << val;
+    ptr += 4;
+  }
+
+  if (ptr != end) {
+    size_t bytes = end - ptr;
+    if (bytes / 4 * 4 != bytes) {
```
Comment (fowles) [inline upb/mini_table.c:999]: `bytes % 4 != 0` is a simpler way of saying this

File: upb/mini_table.c [hunk 4]
```diff
@@ -884,34 +962,71 @@ upb_MiniTable* upb_MiniTable_BuildMapEntry(upb_FieldType key_type,
...
+      upb_MtDecoder_ErrorFormat(&decoder, "Bytes should be a multiple of 4");
+      UPB_UNREACHABLE();
+    }
+    table->values = upb_Arena_Malloc(arena, end - ptr);
+    upb_MtDecoder_CheckOutOfMemory(&decoder, table);
+    memcpy((void*)table->values, ptr, end - ptr);
```
Comment (fowles) [inline upb/mini_table.c:1005]: you shouldn't need this `(void*)` cast I think

File: upb/msg_internal.h [hunk 5]
```diff
@@ -72,20 +72,25 @@ typedef struct {
...
   kUpb_FieldMode_Map = 0,
   kUpb_FieldMode_Array = 1,
   kUpb_FieldMode_Scalar = 2,
 
-  kUpb_FieldMode_Mask = 3, /* Mask to isolate the mode from upb_FieldRep. */
 } upb_FieldMode;
 
+#define kUpb_FieldMode_Mask                        \
+  3 /* Mask to isolate the mode from upb_FieldRep. \
+     */
```
Comment (fowles) [inline upb/msg_internal.h:92]: Is there some reason to have this macro expand to include a comment?

File: upbc/protoc-gen-upb.cc [hunk 6]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+    kEnum = 2,
+    kMask = 3,
+  };
+
+  static upb_MiniTable_Sub PackSub(const char* data, SubTag tag);
+  static bool IsNull(upb_MiniTable_Sub sub);
+  void SetSubTableStrings();
+  upb_MiniTable_Sub PackSubForField(const protobuf::FieldDescriptor* f,
+                                    const upb_MiniTable_Field* mt_f);
+  const char* AllocStr(const std::string& str);
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:389]: absl::string_view

File: upbc/protoc-gen-upb.cc [hunk 7]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+
+ private:
+  typedef absl::flat_hash_map<const protobuf::Descriptor*, upb_MiniTable*>
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:392]: prefer `using Foo = Bar` to typedef in new code.  It has a more consistent syntax and works with templates

File: upbc/protoc-gen-upb.cc [hunk 8]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+  return &it->second;
+}
+
+void FilePlatformLayout::ResolveIntraFileReferences() {
+  // This properly resolves references within a file, in order to set any
+  // necessary flags (eg. is a map).
+  for (const auto& pair : table_map_) {
+    upb_MiniTable* mt = pair.second;
+    // First we properly resolve for defs within the file.
+    for (const auto& f : FieldNumberOrder(pair.first)) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:423]: looks like you are taking a `&` to a `*`.  If you can make this a `const auto*` that would be cleaner

File: upbc/protoc-gen-upb.cc [hunk 9]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+      if (f->message_type() && f->message_type()->file() == f->file()) {
+        upb_MiniTable_Field* mt_f = const_cast<upb_MiniTable_Field*>(
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:425]: please add a comment stating why this `const_cast` is safe

File: upbc/protoc-gen-upb.cc [hunk 10]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+      // never alter the mini-table.
+    }
+  }
+}
+
+upb_MiniTable_Sub FilePlatformLayout::PackSub(const char* data, SubTag tag) {
+  uintptr_t val = reinterpret_cast<uintptr_t>(data);
+  assert((val & kMask) == 0);
+  upb_MiniTable_Sub sub;
+  sub.submsg = reinterpret_cast<upb_MiniTable*>(val | tag);
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:440]: you need to keep submsg as a `uintptr_t` if you wish to avoid UB here

File: upbc/protoc-gen-upb.cc [hunk 11]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+  const char* str = reinterpret_cast<const char*>(as_int & ~SubTag::kMask);
+  switch (as_int & SubTag::kMask) {
+    case SubTag::kMessage:
+      return absl::Substitute("{.submsg = &$0}", str);
+    case SubTag::kEnum:
+      return absl::Substitute("{.subenum = &$0}", str);
+    default:
+      return std::string("{.submsg = NULL}");
+  }
+  return std::string("ERROR in WriteSub");
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:459]: GetSub?

File: upbc/protoc-gen-upb.cc [hunk 12]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+}
+
+void FilePlatformLayout::SetSubTableStrings() {
+  for (const auto& pair : table_map_) {
+    upb_MiniTable* mt = pair.second;
+    for (const auto& f : FieldNumberOrder(pair.first)) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:465]: `const auto*`

File: upbc/protoc-gen-upb.cc [hunk 13]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+      upb_MiniTable_Field* mt_f = const_cast<upb_MiniTable_Field*>(
+          upb_MiniTable_FindFieldByNumber(mt, f->number()));
+      assert(mt_f);
+      upb_MiniTable_Sub sub = PackSubForField(f, mt_f);
+      if (IsNull(sub)) continue;
+      *const_cast<upb_MiniTable_Sub*>(&mt->subs[mt_f->submsg_index]) = sub;
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:471]: more `const_cast` comments required

File: upbc/protoc-gen-upb.cc [hunk 14]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+void FilePlatformLayout::BuildMiniTables(const protobuf::FileDescriptor* fd) {
+  for (const auto& m : SortedMessages(fd)) {
+    table_map_[m] = MakeMiniTable(m);
+  }
+  ResolveIntraFileReferences();
+  SetSubTableStrings();
+}
+
+void FilePlatformLayout::BuildExtensions(const protobuf::FileDescriptor* fd) {
+  auto sorted = SortedExtensions(fd);
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:505]: I would prefer a type for this

File: upbc/protoc-gen-upb.cc [hunk 15]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+  upb::Status status;
+  for (const auto& f : sorted) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:507]: `const auto*`

File: upbc/protoc-gen-upb.cc [hunk 16]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+    bool ok = upb_MiniTable_BuildExtension(e.data().data(), e.data().size(),
+                                           &ext, sub, status.ptr());
+    if (!ok) {
+      fprintf(stderr, "Error building mini-table: %s\n",
+              status.error_message());
+    }
+    ext.extendee = reinterpret_cast<const upb_MiniTable*>(
+        AllocStr(MessageInit(f->containing_type())));
+    ext.sub = PackSubForField(f, &ext.field);
+    ABSL_ASSERT(ok);
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:523]: can you move this up to line 520?

File: upbc/protoc-gen-upb.cc [hunk 17]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+    upb::MtDataEncoder e;
+    e.StartMessage(0);
+    e.PutField(static_cast<upb_FieldType>(f->type()), f->number(),
+               GetFieldModifiers(f));
+    upb_MiniTable_Extension& ext = extension_map_[f];
+    upb_MiniTable_Sub sub;
+    bool ok = upb_MiniTable_BuildExtension(e.data().data(), e.data().size(),
+                                           &ext, sub, status.ptr());
+    if (!ok) {
+      fprintf(stderr, "Error building mini-table: %s\n",
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:517]: not sure how much this buys you given the assert.  Also, the fact that it blindly continues is a bit troubling

File: upbc/protoc-gen-upb.cc [hunk 18]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+  } else {
+    return MakeRegularMiniTable(m);
+  }
+}
+
+upb_MiniTable* FilePlatformLayout::MakeRegularMiniTable(
+    const protobuf::Descriptor* m) {
+  upb::MtDataEncoder e;
+  e.StartMessage(GetMessageModifiers(m));
+  for (const auto& f : FieldNumberOrder(m)) {
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:548]: `const auto*`

File: upbc/protoc-gen-upb.cc [hunk 19]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+    e.PutField(static_cast<upb_FieldType>(f->type()), f->number(),
+               GetFieldModifiers(f));
+  }
+  for (int i = 0; i < m->real_oneof_decl_count(); i++) {
+    const protobuf::OneofDescriptor* oneof = m->oneof_decl(i);
+    e.StartOneof();
+    for (int j = 0; j < oneof->field_count(); j++) {
+      const protobuf::FieldDescriptor* f = oneof->field(j);
+      e.PutOneofField(f->number());
+    }
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:558]: do you need an `e.EndOneof()` or something?

File: upbc/protoc-gen-upb.cc [hunk 20]
```diff
@@ -369,6 +329,346 @@ std::string CTypeConst(const protobuf::FieldDescriptor* field) {
...
+  }
+  const auto& str = e.data();
```
Comment (fowles) [inline upbc/protoc-gen-upb.cc:560]: please provide the type here.

File: upb/mini_table.c [hunk 21]
```diff
@@ -110,6 +110,14 @@ char upb_FromBase92(uint8_t ch) {
   return kUpb_FromBase92[ch - ' '];
 }
 
+bool upb_IsTypePackable(upb_FieldType type) {
+  static const unsigned kPackableTypes =
+      -1U & ~(1 << kUpb_FieldType_String) & ~(1U << kUpb_FieldType_Bytes) &
+      ~(1U << kUpb_FieldType_Message) & ~(1U << kUpb_FieldType_Group);
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/mini_table.c:116]: Done.

File: upb/mini_table.c [hunk 22]
```diff
@@ -152,15 +160,26 @@ static char* upb_MtDataEncoder_PutBase92Varint(upb_MtDataEncoder* e, char* ptr,
...
                                      uint64_t msg_mod) {
   upb_MtDataEncoderInternal* in = upb_MtDataEncoder_GetInternal(e, ptr);
   in->msg_mod = msg_mod;
   in->last_field_num = 0;
   in->oneof_state = kUpb_OneofState_NotStarted;
-  return ptr;
+  return upb_MtDataEncoder_PutModifier(e, ptr, msg_mod);
 }
 
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/mini_table.c:181]: Done.

File: upb/mini_table.c [hunk 23]
```diff
@@ -197,33 +215,36 @@ char* upb_MtDataEncoder_PutField(upb_MtDataEncoder* e, char* ptr,
...
+    type = kUpb_FieldType_Int32;
+  }
+
   int encoded_type = kUpb_TypeToEncoded[type];
   if (field_mod & kUpb_FieldModifier_IsRepeated) {
     // Repeated fields shift the type number up (unlike other modifiers which
     // are bit flags).
     encoded_type += kUpb_EncodedType_RepeatedBase;
+
+    bool field_is_packed = field_mod & kUpb_FieldModifier_IsPacked;
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/mini_table.c:232]: Done.

File: upb/mini_table.c [hunk 24]
```diff
@@ -351,10 +372,41 @@ static const char* upb_MiniTable_DecodeBase92Varint(upb_MtDecoder* d,
   }
 }
 
-static bool upb_MiniTable_HasSub(char type, uint64_t msg_modifiers) {
-  return type == kUpb_EncodedType_Message || type == kUpb_EncodedType_Group ||
-         (type == kUpb_EncodedType_Enum &&
-          (msg_modifiers & kUpb_MessageModifier_HasClosedEnums));
+static bool upb_MiniTable_HasSub(upb_MiniTable_Field* field,
+                                 uint64_t msg_modifiers) {
+  if (field->descriptortype == kUpb_FieldType_Message ||
```
[the remark this answers is not in the document]
Comment (haberman) [reply upb/mini_table.c:377]: Done.

File: upb/mini_table.c (hunk 1 above)
Comment (haberman) [reply upb/mini_table.c:477]: Done.

File: upb/mini_table.c (hunk 2 above)
Comment (haberman) [reply upb/mini_table.c:565]: For the compiler to generate platform-independent code, a single binary must be capable of calculating both 32 and 64-bit layouts at runtime. That requires the binary to contain both tables. This is how the compiler can generate `UPB_SIZE(x, y)` for the field offsets.

File: upb/mini_table.c (hunk 3 above)
Comment (haberman) [reply upb/mini_table.c:999]: Done.

File: upb/mini_table.c (hunk 4 above)
Comment (haberman) [reply upb/mini_table.c:1005]: It's casting away const.

File: upb/msg_internal.h (hunk 5 above)
Comment (haberman) [reply upb/msg_internal.h:92]: Fixed.

File: upbc/protoc-gen-upb.cc (hunk 6 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:389]: I think that might be less clear here, as the NULL-terminated nature of the returned string is critical. It is a C string that will be stored by pointer only.

File: upbc/protoc-gen-upb.cc (hunk 7 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:392]: Done.

File: upbc/protoc-gen-upb.cc (hunk 8 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:423]: Done.

File: upbc/protoc-gen-upb.cc (hunk 10 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:440]: Are you sure? Point 5 here suggests to me that I can count on the pointer to safely round-trip: https://en.cppreference.com/w/cpp/language/reinterpret_cast

It's true that `sub.submsg` has a greater alignment requirement, however I know a priori that the arena aligns all allocations to 16, so the pointer should satisfy the alignment requirement here.

File: upbc/protoc-gen-upb.cc (hunk 11 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:459]: Done.

File: upbc/protoc-gen-upb.cc (hunk 12 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:465]: Done.

File: upbc/protoc-gen-upb.cc (hunk 9 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:425]: Done.

File: upbc/protoc-gen-upb.cc (hunk 13 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:471]: Done.

File: upbc/protoc-gen-upb.cc (hunk 14 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:505]: Done.

File: upbc/protoc-gen-upb.cc (hunk 15 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:507]: Done.

File: upbc/protoc-gen-upb.cc (hunk 16 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:523]: Done.

File: upbc/protoc-gen-upb.cc (hunk 17 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:517]: Bazel makes it very hard to use an asserts-enabled build of a code generator during the build. I'm sort of biding my time until we have ABSL logging and can use `CHECK()`.

File: upbc/protoc-gen-upb.cc (hunk 18 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:548]: Done.

File: upbc/protoc-gen-upb.cc (hunk 19 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:558]: Nope, it's not currently required. Oneofs have separators between them, but no termination.

File: upbc/protoc-gen-upb.cc (hunk 20 above)
Comment (haberman) [reply upbc/protoc-gen-upb.cc:560]: Done.

**Comment by haberman** [comment]
PTAL.

File: upbc/protoc-gen-upb.cc (hunk 6 above)
Comment (fowles) [reply upbc/protoc-gen-upb.cc:389]: I meant for the parameter, not the return type

File: upbc/protoc-gen-upb.cc (hunk 17 above)
Comment (fowles) [reply upbc/protoc-gen-upb.cc:520]: can you add a TODO?

**Review by fowles — APPROVED** [review]

**Pull request merged**